### PR TITLE
Brave instrumentation for Akka Streams

### PR DESCRIPTION
**Pull request by jeqo**
Work in progress: adding Akka Streams components that start/finish/wrap flows.

The main goal of this connectors is to measure latency on internal Akka Streams applications. E.g:

Stream application: JmsSource -> Parsing -> Sending to Kafka -> Close

Should create a trace:

```
+-------- main ---------------------------------------+
      +-parsing-+ 
                            +---send-to-kafka-----+
```

Missing tasks:
- [x] Testing
- [x] Examples
- [x] Docs

### CODE REVIEW

File: alpakka-brave/src/test/java/no/sysco/middleware/alpakka/brave/javadsl/examples/HelloWorld.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package no.sysco.middleware.alpakka.brave.javadsl.examples;
+
+public class HelloWorld {
+  //public static void main(String[] args) {
+  //  public static void main(String[] args) {
+  //  final ActorSystem system = ActorSystem.create();
+  //  final ActorMaterializer materializer = ActorMaterializer.create(system);
+  //
+  //  final Sender sender = URLConnectionSender.create("http://zipkin-zipkin.paas-test.statnett.no/api/v2/spans");
```
Comment (chromy96) [inline alpakka-brave/src/test/java/no/sysco/middleware/alpakka/brave/javadsl/examples/HelloWorld.java]: Get rid of URL :)

File: alpakka-brave/src/test/java/no/sysco/middleware/alpakka/brave/javadsl/examples/HelloWorld.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package no.sysco.middleware.alpakka.brave.javadsl.examples;
+
+public class HelloWorld {
```
Comment (chromy96) [inline alpakka-brave/src/test/java/no/sysco/middleware/alpakka/brave/javadsl/examples/HelloWorld.java]: Should this example be commented out? If so, maybe you can get rid of it?

File: alpakka-brave/src/main/java/no/sysco/middleware/alpakka/brave/javadsl/Brave.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  public static <T> Graph<FlowShape<T, Pair<T, Span>>, NotUsed> startSpan(final Tracing tracing,
+                                                                          final String spanName) {
+    return Flow.fromGraph(new BraveStartSpanFlowStage<>(tracing, spanName));
+  }
+
+  public static <T> Graph<FlowShape<Pair<T, Span>, T>, NotUsed> finishSpan(final Tracing tracing) {
+    return Flow.fromGraph(new BraveFinishSpanFlowStage<>(tracing));
+  }
+
+  public static <A, B, Mat> Graph<FlowShape<Pair<A, Span>, Pair<B, Span>>, Mat> spanFlow(final Tracing tracing,
```
Comment (chromy96) [inline alpakka-brave/src/main/java/no/sysco/middleware/alpakka/brave/javadsl/Brave.java]: Naming seems little inconsistent: startSpan, finishSpan, spanFlow.
Maybe flowSpan would be more consistent with the other two methods. What do you think?

File: alpakka-brave/src/test/java/no/sysco/middleware/alpakka/brave/javadsl/examples/HelloWorld.java (hunk 2 above)
Comment (jeqo) [reply alpakka-brave/src/test/java/no/sysco/middleware/alpakka/brave/javadsl/examples/HelloWorld.java]: We can remove it in the meantime, I will add some proper tests.

**Comment by jeqo** [comment]
@chromy96 I have deleted comments to remove cache of comment. I think is clean now.

File: alpakka-brave/src/main/java/no/sysco/middleware/alpakka/brave/javadsl/BraveFinishSpanFlowStage.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  public FlowShape<Pair<T, Span>, T> shape() {
+    return new FlowShape<>(in, out);
+  }
+
+  public GraphStageLogic createLogic(Attributes inheritedAttributes) {
+    return new GraphStageLogic(shape()) {
+
+      {
+        final Tracer tracer = tracing.tracer();
```
Comment (timurgen) [inline alpakka-brave/src/main/java/no/sysco/middleware/alpakka/brave/javadsl/BraveFinishSpanFlowStage.java]: Seems like this is unused variable tracer here

**Review by timurgen — CHANGES_REQUESTED** [review]

**Review by timurgen — APPROVED** [review]

**Comment by jeqo** [comment]
Add some more tests, I think could be enough for a initial release.

**Comment by jeqo** [comment]
Merging this. If there is any missing improvement, please create an issue.

**Pull request merged**